odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,497 @@
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+    >>> linear_deform(template, displacement_field)
+    array([ 0.,  0.,  1.,  1.,  0.])
+    """
+    assert template in displacement[0].space
+
+    image_pts = template.space.grid.points()
+    image_pts += np.asarray(displacement).T
+
+    return template.interpolation(image_pts.T, bounds_check=False)
+
+
+def deform_grad(grad_f, displacement):
+    """Compute the deformation of the template gradient.
+
+    The template gradient ``grad(I)`` after deformation with displacement
+    field ``v``: grad(I)(Id + v)
+
+    Parameters
+    ----------
+    grad_f: `ProductSpaceVector`
+        Gradient of the template, i.e. a vector field on the
+        image domain
+    displacement : `ProductSpaceVector`
+        The vector field (displacement field) used in the linearized
+        deformation.
+    Examples
+    --------
+    >>> import odl
+
+    Define ``gradOp`` as the gradient operator, let it act on ``template``.
+
+    >>> space = odl.uniform_discr(0, 1, 5)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> gradOp = odl.Gradient(template.space, method='forward')
+    >>> template_grad = gradOp(template)
+    >>> template_grad
+    ProductSpace(uniform_discr(0.0, 1.0, 5), 1).element([
+        [0.0, 5.0, -5.0, 0.0, 0.0]
+    ])
+
+    Define a displacement field ``displacement_field``.
+
+    >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+
+    Calculate linearized deformation of the ``template_grad`` (gradient of
+    the template) using the displacement field ``displacement_field``.
+
+    >>> deform_grad(template_grad, displacement_field)
+    ProductSpace(uniform_discr(0.0, 1.0, 5), 1).element([
+        [0.0, 5.0, -5.0, -5.0, 0.0]
+    ])
+    """
+    temp = [linear_deform(gf, displacement) for gf in grad_f]
+    return displacement.space.element(temp)
+
+
+class LinDeforFixedTempOp(Operator):
+    """Operator mapping a displacement field to corresponding deformed template.
+
+    The operator maps a displacement field to the corresponding
+    deformed fixed template ``I``. The linearized deformation is given by
+    the left action defined through the pushforward, so
+
+        v --> I(Id + v)
+
+    where ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement field:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain of the target.
+    """
+    def __init__(self, template):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template that is to be deformed by the displacement field.
+
+        Examples
+        --------
+        Deform a 1D function ``template`` that has value ``1`` at 3:rd sample
```

Comment:
As stated above. This is mostly a **what** style documentation, that is quite clear from simply reading the code below. Users would be more interested in a **why**, and in some cases **how**, style documentation.

Here, you may want to start with the first line, simply to introduce the problem.

"Deform a 1D function ``template`` using a displacement field."

and then explain more in depth:

"Since all points except the 4:th are sampled with 0 offset, they should have the same value as before the deformation. The 4:th point is shifted 0.2 to the left, and hence should get the same value as the 3:rd point"